L/LightGBM/build_tarballs.jl

```diff
@@ -17,27 +21,49 @@ cd $WORKSPACE/srcdir/LightGBM
 git submodule update --init --depth=1
 git submodule update --checkout --depth=1
 
+FLAGS=()
+cmake_extra_args=""
 
-if [[ $target == *"apple-darwin"* ]]; then
-  cmake_extra_args="-DAPPLE=1 -DAPPLE_OUTPUT_DYLIB=1"
+if [[ "${bb_full_target}" == *"apple-darwin"* ]]; then
+  cmake_extra_args+="-DAPPLE=1 -DAPPLE_OUTPUT_DYLIB=1 "
 fi
 
-FLAGS=()
-
-if [[ "${target}" == *-mingw* ]]; then
-  # Parches windows
-  for p in $WORKSPACE/srcdir/patches/windows/*.patch; do
-    atomic_patch -p1 "${p}"
-  done
-  cmake_extra_args="-DWIN32=1 -DMINGW=1"
+if [[ "${bb_full_target}" == *-mingw* ]]; then
+  cmake_extra_args+="-DWIN32=1 -DMINGW=1 "
   FLAGS+=(LDFLAGS="-no-undefined")
 fi
 
-cmake \
-  -DCMAKE_INSTALL_PREFIX=$prefix\
-  -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
-  -DCMAKE_BUILD_TYPE=Release \
-  $cmake_extra_args
+if [[ "${bb_full_target}" == *-linux* ]]; then
+  cmake_extra_args+="-DUSE_GPU=1 "
+fi
+
+if  [[ ("${bb_full_target}" == *-cuda*) && ("${bb_full_target}" != *-cuda+none*) ]]; then
+  # nvcc writes to /tmp, which is a small tmpfs in our sandbox.
+  # make it use the workspace instead
+  export TMPDIR=${WORKSPACE}/tmpdir
+  mkdir ${TMPDIR}
+  export CUDA_HOME=${WORKSPACE}/destdir/cuda
+  export PATH=$PATH:$CUDA_HOME/bin/
+
+  ln -s $CUDA_HOME/lib $CUDA_HOME/lib64
+
+  cmake_extra_args+="\
+    -DUSE_CUDA=1 \
+    -DCUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME \
+    -DCUDA_TOOLKIT_INCLUDE=$CUDA_HOME/include \
+    -DCUDA_CUDART_LIBRARY=$CUDA_HOME/lib64/libcudart.$dlext "
+fi
+configure () {
+  cmake . \
+    -DCMAKE_INSTALL_PREFIX=$prefix\
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_BUILD_TYPE=Release \
+    $cmake_extra_args
+}
+
+# Need to call twice due to FindCUDA issues in cmake
+configure || configure
```

Comment:
What issues? I haven't seen problems with CMake's CUDA support requiring two configuration steps to solve.

Reply:
it has to do with cmake finding binaries and libs of cuda https://github.com/JuliaPackaging/Yggdrasil/pull/4554#discussion_r831155780.
CUDA build throw this error the first time cmake is called https://buildkite.com/julialang/yggdrasil/builds/10093#018f3bfb-663a-4a64-a57f-d5148ee1dbdb/660-912
```
[01:16:29] CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
[01:16:29]   Could NOT find CUDA (missing: CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) (found
[01:16:29]   suitable version "11.5", minimum required is "11.0")
[01:16:29] Call Stack (most recent call first):
[01:16:29]   /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
[01:16:29]   /usr/share/cmake/Modules/FindCUDA.cmake:1264 (find_package_handle_standard_args)
[01:16:29]   CMakeLists.txt:201 (find_package)
```

Reply:
They shouldn't be using the FindCUDA module via `find_package(CUDA)`, since it has been deprecated since CMake 3.10: https://cmake.org/cmake/help/latest/module/FindCUDA.html. Instead they just need to enable the language for CUDA and CMake should take care of the rest of the setup for the base CUDA libraries.

Reply:
thanks, let me see if I can get the change upstream, do you think we can merge this as is or this need a patch to their CMakeLists.txt?

Reply:
Should this be ok to merge?

Reply:
Yea, I don't think this is blocking for it, it is just very odd it takes two tries to pass... but if it works :man_shrugging: